Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -6414,6 +6417,196 @@ bool Document::fullscreen_enabled() const
+            }
+        }
+
+        // 15. For each descendantDoc in descendantDocs:
+        for (auto& descendant_doc : descendant_docs) {
+            // 1. Append (fullscreenchange, descendantDoc’s fullscreen element) to descendantDoc’s list of pending fullscreen events.
+            descendant_doc->append_pending_fullscreen_change(PendingFullscreenEvent::Type::Change, *descendant_doc->fullscreen_element());
+            // 2. Unfullscreen descendantDoc.
+            GC::RootVector<GC::Ref<Element>, 8> fullscreen_elements(realm.heap());
+            for (auto const& element : descendant_doc->top_layer_elements()) {
+                if (element->is_fullscreen_element())
+                    fullscreen_elements.append(element);
+            }
+
+            for (auto& el : fullscreen_elements) {
+                descendant_doc->unfullscreen_element(el);
+            }
+        }
+
+        // NOTE: The order in which documents are unfullscreened is not observable, because run the fullscreen steps is invoked in tree order.
+        // 16. Resolve promise with undefined.
+        WebIDL::resolve_promise(realm, promise, JS::js_undefined());
+    }));
+    return promise;
+}
+
+// https://fullscreen.spec.whatwg.org/#simple-fullscreen-document
+bool Document::is_simple_fullscreen_document() const
+{
+    // A document is said to be a simple fullscreen document if there is exactly one element in its top layer that has its fullscreen flag set.
+    u32 total = 0;
+    for (auto const& element : top_layer_elements()) {
+        if (element->is_fullscreen_element())
+            ++total;
+
+        if (total > 1)
+            return false;
+    }
+    return total == 1;
+}
+
+// https://fullscreen.spec.whatwg.org/#collect-documents-to-unfullscreen
+GC::RootVector<GC::Ref<Document>> Document::collect_documents_to_unfullscreen() const
+{
+    // 1. Let docs be an ordered set consisting of doc.
+    GC::RootVector<GC::Ref<Document>> docs { heap() };
+    docs.append(GC::Ref { *const_cast<Document*>(this) });
+    // 2. While true:
+    for (;;) {
+        //  1. Let lastDoc be docs’s last document.
+        auto last_doc = docs.last();
+        //  2. Assert: lastDoc’s fullscreen element is not null.
+        VERIFY(last_doc->fullscreen_element());
+        //  3. If lastDoc is not a simple fullscreen document, break.
+        if (!last_doc->is_simple_fullscreen_document())
+            break;
+        //  4. Let container be lastDoc’s node navigable’s container.
+        // Note on spec: It doesn't first check if `node navigable` is null.
+        auto container = last_doc->navigable() ? last_doc->navigable()->container() : nullptr;
+        //  5. If container is null, then break.
+        if (!container)
+            break;
+        //  6. If container’s iframe fullscreen flag is set, break.
+        if (HTML::HTMLIFrameElement* iframe_element = as<HTML::HTMLIFrameElement>(container.ptr()); iframe_element) {
```

Comment:
This `as<HTML::HTMLIFrameElement>` will panic if `container.ptr()` is not the right type, so you can forego the `iframe_element` condition and add the `iframe_fullscreen_flag()` condition instead.